### ISSUE TITLE
Parallel build not complete

### ISSUE THREAD

**Issue by bdbaddog** [SCons/scons#1636]
This issue was originally created at: 2007-04-13 09:09:16.
This issue was reported by: `issues@scons`.
issues@scons said at 2007-04-13 09:09:16
>I am producing a build system based on SCons and have recently made it default
>to using 4 threads. One of the guys has found a (repeatable) situation where the
>build is not complete when scons is run the first time it needs to be run again
>to complete the build. By contrast, when run with a single thread, the build
>completes correctly the first time.
> 
>To summarise the target that he is trying to build:
>1) A bunch of C-files are compiled to produce a bunch of .obj-files
>2) The .obj-files are linked to produce a .map-file and a
>(cpumanf_pre_cksum).out-file (I don't know why the .map is included as an
>explicit output of the tool; the tool was written by someone who is no longer at
>the company)
>3) An application (addcheck.exe) is used to take the .out-file, add a checksum
>and produce another (cpumanf).out-file. This is done using an Action inside a
>Builder.
>4) This final .out-file is installed in another directory
> 
>When a single C-file is changed, the parallel build will stop after step 2 the
>first time it is run, completing only if run again. The serial build will run
>through all 4 steps the first time.
> 
>I've turned on taskmaster tracing and got the following for the parallel build
>(trimed to only the end of it):
> 
>==== Start parallel build trace ====
>Taskmaster:
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.map':
>children:
>   
>['C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\common.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\intmath.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\mezz_readback.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\timebase.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_main.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\cpumanf_link.cmd',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\manf\\build\\c6416\\manf.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_McBSP.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_TCP.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_VCP.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_cvectors.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_edma.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_initialise.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_intr.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_intrinsics.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_kernel.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_logging.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_mezz.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_prefetch.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_system.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_timebase.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_trace.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_transport.obj']
>    waiting on unfinished children:
>    ['C:\\lte\\ubibs_lte_work\\lte_dsp_app\\manf\\build\\c6416\\manf.obj']
>Taskmaster: 'domain\\dsp\\addcheck\\addcheck.exe': children:
>    ['domain\\dsp\\addcheck\\addcheck.obj']
>    waiting on unstarted children:
>    ['domain\\dsp\\addcheck\\addcheck.obj']
>Taskmaster: 'domain\\dsp\\addcheck\\addcheck.obj': children:
>    ['domain\\dsp\\addcheck\\addcheck.c']
>    evaluating domain\dsp\addcheck\addcheck.obj
>Taskmaster: 'domain\\dsp\\addcheck\\addcheck.exe': children:
>    ['domain\\dsp\\addcheck\\addcheck.obj']
>    evaluating domain\dsp\addcheck\addcheck.exe
>Taskmaster: 'domain\\dsp\\addcheck\\addcheck.exe': already handled (up_to_date)
>Taskmaster:
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf.out':
>children:
>   
>['C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.map',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.out',
>'domain\\dsp\\addcheck\\addcheck.exe']
>    waiting on unfinished children:
>   
>['C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.map',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.out']
>Taskmaster: 'build\\target\\cpumanf.out': children:
>   
>['C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf.out']
>    waiting on unfinished children:
>   
>['C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf.out']
>Taskmaster: 'cpumanf': children:
>   
>['C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\common.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\intmath.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\mezz_readback.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\timebase.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_main.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\manf\\build\\c6416\\manf.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_McBSP.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_TCP.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_VCP.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_cvectors.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_edma.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_initialise.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_intr.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_intrinsics.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_kernel.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_logging.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_mezz.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_prefetch.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_system.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_timebase.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_trace.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_transport.obj',
>'build\\target\\cpumanf.out']
>    waiting on unfinished children:
>    ['C:\\lte\\ubibs_lte_work\\lte_d	sp_app\\manf\\build\\c6416\\manf.obj',
>'build\\target\\cpumanf.out']
>Taskmaster:
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.map':
>children:
>   
>['C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\common.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\intmath.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\mezz_readback.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\timebase.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_main.obj',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\cpumanf_link.cmd',
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\manf\\build\\c6416\\manf.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_McBSP.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_TCP.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_VCP.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_cvectors.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_edma.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_initialise.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_intr.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_intrinsics.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_kernel.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_logging.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_mezz.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_prefetch.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_system.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_timebase.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_trace.obj',
>'C:\\lte\\ubibs_lte_work\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_transport.obj']
>    evaluating
>C:\lte\ubibs_lte_work\lte_dsp_app\main\cpumanf\build\c6416\cpumanf_pre_cksum.map
>Building cpumanf_pre_cksum.out and cpumanf_pre_cksum.map...
>Taskmaster:
>'C:\\lte\\ubibs_lte_work\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.out':
>already handled (executing)
>scons: done building targets.
>==== End parallel build trace ====
> 
>For the serial build, I got (again trimmed):
> 
>==== Start serial build trace ====
>Taskmaster: 'build\\target\\cpumanf.out': children:
>    ['C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf.out']
>    waiting on unstarted children:
>    ['C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf.out']
>Taskmaster:
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf.out':
>children:
>   
>['C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.map',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.out',
>'domain\\dsp\\addcheck\\addcheck.exe']
>    waiting on unstarted children:
>   
>['C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.map',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.out',
>'domain\\dsp\\addcheck\\addcheck.exe']
>Taskmaster:
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.out':
>children:
>   
>['C:\\lte\\UBIBS_~2\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\common.obj',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\intmath.obj',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\mezz_readback.obj',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\timebase.obj',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_main.obj',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\cpumanf_link.cmd',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\manf\\build\\c6416\\manf.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_McBSP.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_TCP.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_VCP.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_cvectors.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_edma.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_initialise.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_intr.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_intrinsics.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_kernel.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_logging.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_mezz.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_prefetch.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_system.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_timebase.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_trace.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_transport.obj']
>    evaluating
>C:\lte\UBIBS_~2\lte_dsp_app\main\cpumanf\build\c6416\cpumanf_pre_cksum.out
>Building cpumanf_pre_cksum.out and cpumanf_pre_cksum.map...
>Taskmaster:
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.map':
>already handled (executed)
>Taskmaster: 'domain\\dsp\\addcheck\\addcheck.exe': children:
>    ['domain\\dsp\\addcheck\\addcheck.obj']
>    waiting on unstarted children:
>    ['domain\\dsp\\addcheck\\addcheck.obj']
>Taskmaster: 'domain\\dsp\\addcheck\\addcheck.obj': children:
>    ['domain\\dsp\\addcheck\\addcheck.c']
>    evaluating domain\dsp\addcheck\addcheck.obj
>Taskmaster: 'domain\\dsp\\addcheck\\addcheck.exe': children:
>    ['domain\\dsp\\addcheck\\addcheck.obj']
>    evaluating domain\dsp\addcheck\addcheck.exe
>Taskmaster:
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf.out':
>children:
>   
>['C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.map',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_pre_cksum.out',
>'domain\\dsp\\addcheck\\addcheck.exe']
>    evaluating C:\lte\UBIBS_~2\lte_dsp_app\main\cpumanf\build\c6416\cpumanf.out
>Copying cpumanf_pre_cksum.out to cpumanf.out with checksum...
>Taskmaster: 'build\\target\\cpumanf.out': children:
>    ['C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf.out']
>    evaluating build\target\cpumanf.out
>Copying cpumanf.out to build\target...
>Taskmaster: 'cpumanf': children:
>   
>['C:\\lte\\UBIBS_~2\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\common.obj',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\intmath.obj',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\mezz_readback.obj',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\common\\build\\c6416\\cpumanf\\timebase.obj',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf_main.obj',
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\manf\\build\\c6416\\manf.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_McBSP.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_TCP.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_VCP.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_cvectors.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_edma.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_initialise.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_intr.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_intrinsics.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_kernel.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_logging.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_mezz.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_prefetch.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_system.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_timebase.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_trace.obj',
>'C:\\lte\\UBIBS_~2\\tm_system\\dspkernel\\build\\c6416\\cpumanf\\k_transport.obj',
>'build\\target\\cpumanf.out']
>    evaluating cpumanf
>Taskmaster: 'domain\\dsp\\addcheck\\addcheck.exe': already handled (up_to_date)
>Taskmaster:
>'C:\\lte\\UBIBS_~2\\lte_dsp_app\\main\\cpumanf\\build\\c6416\\cpumanf.out':
>already handled (executed)
>Taskmaster: 'build\\target\\cpumanf.out': already handled (executed)
>Taskmaster: 'cpumanf': already handled (executed)
>scons: done building targets.
>==== End serial build trace ====
> 
>Please let me know if there is any additional imformation that I can provide.
> 
>Steve.

stevenknight said at 2007-04-13 09:52:44
>Is this a new problem?  Does it happen only with 0.96.96, or did it also seem to
>happen with any earlier versions you were using?
> 
>Actually, it sounds very much like a problem that was fixed in 0.96.96.  Since I
>only released 0.96.96 late last night, have you actually downloaded it already
>and confirmed the problem with 0.96.96, or is the Version field in this bug
>report mistaken?
> 
>If you're able to reproduce it under 0.96.96, we'll likely need to take a look
>at the relevant parts of the configuration.  Is there any chance you can strip
>your configuration down to a subset that demonstrates the problem?
> 
>        --SK

issues@scons said at 2007-04-15 02:12:49
>Hi SK,
> 
>We've been working with 0.96.95 - with "fixed" in the installer file name, IIRC
>- and I saw the bug you refered to so I tried i) applying the patch to 0.96.95
>and ii) downloading and using 0.96.96, but neither fixed the problem.
> 
>I can certainly try to come up with a minimal configuration that exhibits the
>problem, but the build system is very complex so it could take me a while. If
>there is any diagnostics I can get you from the real system in the meantime,
>please let me know.
> 
>Cheers,
> 
>Steve.

stevenknight said at 2007-04-16 16:50:01
>Hi Steve--
> 
>Here's additional info you could capture that might be useful without having to
>modify the configuration:
> 
>    Run with --tree=all each time you build.  There shouldn't be any differences
>between the dependency graphs, but if there are, then it would provide a good
>clue to what's going on.
> 
>    Run with --taskmastertrace without using -j (i.e., when it succesffully
>builds everything) so we can compare the difference between how it visits the
>Nodes.  It would help if you could re-post the taskmastertrace output you
>already posted as an attachment, not pasted into the comment, so that it can be
>downloaded and compared.  (Trying to look at that output in a browser is a pain
>because the browser wraps text and does other things to try to make the
>formatting look pretty.)
> 
>Thanks,
> 
>        --SK

issues@scons said at 2007-04-17 08:15:04
>Created an attachment (id=174)
>log of "scons cpumanf -j 4 --tree=all --taskmastertrace=-" for the attempt

issues@scons said at 2007-04-17 08:16:36
>Created an attachment (id=175)
>log of "scons cpumanf -j 4 --tree=all --taskmastertrace=-" for the second attempt

issues@scons said at 2007-04-17 08:17:56
>Created an attachment (id=176)
>log of "scons cpumanf --tree=all --taskmastertrace=-" (which succeeds in one attempt)

issues@scons said at 2007-04-17 08:22:42
>SK,
> 
>I've added the attachments that you requested and notice a couple of things:
> - The tree in the second attempt to build using "-j 4" (the one that completes
>the build) matches the the one produced without the -j option
> - The "-j 4" build that doesn't complete does not produce a dependency tree in
>the output
> 
>I spent some time trying to come up with a simpler configuration that exhibits
>the same problem but haven't managed to do so as yet.
> 
>Thanks for your help in looking at this,
> 
>Steve. 

stevenknight said at 2007-04-17 15:55:53
>Hi steve--
> 
>Double-checking a few things:
> 
>1)  You're not globbing on-disk for any files, are you?  It doesn't seem like
>it, but I want to double-check...
> 
>2)  Which of these nodes in the tree are Aliases?  Only "cpumanf?"  Or is
>"build\target\cpumanf.out" an Alias as well?  Are there any other Aliases in the
>build, perhaps some that match these names?
> 
>Okay, with those out of the way, this problem does look very similar to the one
>that was fixed.  The pass1 output suggests that what's happening is the
>reference count in the Taskmaster is getting messaged up because
>cpumanf_pre_cksum.{out,map} both get built by the same command.  What's supposed
>to happen is that the reference count on build\c6416\cpumanf.out should then go
>to 0 and get put back on the list, but it doesn't, and the build stops in its
>tracks.
> 
>One possibility is that the reference count is getting decremented *past* 0
>because it's not initialized correctly.  You can test for this by changing the
>following line in Taskmaster.py (should be on or about line 327):
> 
>            if p.ref_count == 0:
> 
>to the following:
> 
>            if p.ref_count <= 0:
> 
>This is in the Task.postprocess() method.  If you try that and the build
>succeeds on the first pass, then the initialization is a problem.  If it still
>fails, then we're not decrementing it enough.
> 
>While that's going on, if you could show me the Builder calls that initialize
>the following targets, it might give a clue as to what's going on:
> 
>    cpumanf    [I think that's an Alias?]
>    build\target\cpumanf.out
>    ...\cpumanf\build\c6416\cpumanf.out
>    ...\cpumanf\build\c6416\cpumanf_pre_cksum.out
>    ...\cpumanf\build\c6416\cpumanf_pre_cksum.map
>    ...\cpumanf\build\c6416\cpumanf_main.obj
>    ...\cpumanf\cpumanf_link.cmd    [not sure if this is a built file?]
> 
>Thanks,
> 
>        --SK

stevenknight said at 2007-04-17 16:00:18
>One more question:  are you using SideEffect() anywhere?
> 
>        --SK

issues@scons said at 2007-04-18 10:28:39
>SK,
> 
>To answer your questions as best I can:
> 
>- No globbing; all files are explicitly named.
> 
>- cpumanf is the only alias in this build; there are *loads* of others in the
>build system as a whole but there aren't any that match these names.
> 
>- build/target/cpumanf.out is not an Alias, it is an Install of
>...\cpumanf\build\c6416\cpumanf.out
> 
>- I'll try the change to Taskmaster tomorrow, when the guy who's experiencing
>the problem is back in and let you know
> 
>- I'm not using SideEffect anywhere
> 
>- Regarding the Builder calls that you ask for, it's not straight forward as
>I've written a set of abstraction classes that allow the developers to write
>simple module definitions that the abstraction classes then combine and insert
>into SCons as appropriate. However, I'll do my best:
> 
>  - cpumanf
>    - This is an Alias - and there is a slight bug in my code as it shouldn't
>alias to all the .obj files. However, it it set up in two stages: Firstly, the
>erroneous one "localEnv.Alias([self.name], [output])" (where self.name is
>"cpumanf" and output is the list of .obj files) and, secondly,
>"localEnv.Alias([dspImage], state.retval)" (where state.retval is
>build/target/cpumanf - as returned from the Install that I described above.
> 
>  - build\target\cpumanf.out
>    - localEnv.Install(state.installDir, cpuOutCksum)(where state.installDir is
>build/target and cpuOutCksum is the output of the Builder used to add a checksum
>to the cpumanf_pre_cksum.out)
> 
>  - ...\cpumanf\build\c6416\cpumanf.out
>    - This uses a Builder and Action and involves the addcheck.exe that you can
>see at the bottom of the dependency tree. It is added to the environment like this:
>    checksumExe = SConscript(os.path.join("..", "domain", "dsp", "addcheck",
>"addcheck.SConscript"))[0]
>    # Add the checksum tool to the environment
>    cksumAction= Action(checksumExe.abspath + ' -q $SOURCE $TARGET', printer)
>    bld = SCons.Builder.Builder(action = cksumAction, suffix = '.out',
>src_suffix = '.out')
>    env['BUILDERS']['AddChecksum'] = bld
>and called like this:
>    localEnv.AddChecksum(target = os.path.join(buildDir, dspImage), source =
>cpuOut) (where buildDir is "build/c6416", dspImage is "cpumanf" and cpuOut is
>the output of the call to Program to build cpumanf_pre_cksum.out)
> 
>  - ...\cpumanf\build\c6416\cpumanf_pre_cksum.out
>    ...\cpumanf\build\c6416\cpumanf_pre_cksum.map
>    - These are added by localEnv.Program(target = programTarget, source =
>state.retval) (where programTarget is "build/c6416/cpumanf_pre_cksum" and
>state.retval is the list of Objects comprising this executable)
> 
>  - ...\cpumanf\build\c6416\cpumanf_main.obj
>    - localEnv.SharedObject(os.path.join(buildDir, f.name)) (where f.name is
>"cpumanf_main.c" and buildDir is as above
> 
>  - ...\cpumanf\cpumanf_link.cmd
>    - No, this is not a built file - it is an input command file to the linker.
>The dependency is added as follows:
>    localEnv.Depends(cpuOut, linkerFile) (where linkerFile will be
>"cpumanf_link.cmd")
> 
>Hope that's answered all your question for the moment - I'll get back to you
>once I've tried the Taskmaster change...
> 
>Thanks a lot,
> 
>Steve.

issues@scons said at 2007-04-19 00:45:03
>Right, I got the guy to try changing the "== 0" to "<= 0" and this fixed the
>problem. I bunged in a print and it was the second "if p.ref_count == 0" in the
>method that had p.ref_count = -1.
> 
>Cheers,
> 
>Steve.

stevenknight said at 2007-04-19 05:20:26
>Thanks, that last result is very helpful.  It confirms that we're counting
>decrementing multiple copies of a file when we've only set the reference count
>to a subset of the instances.
> 
>Now I have to isolate a test case that demonstrates this problem so we can make
>sure that whatever the final fix is works and stays fixed in the future.  I'm
>going to try to create a case based on the previous description, but I'll ask if
>I need more information.
> 
>Most likely, it's because of the duplication of the object list in the 'cpumanf'
>Alias and then for both cpumanf_pre_cksum.out and cpumanf_pre_cksum.map, or some
>combination or subset of those three.  Just curious:  You mentioned that the
>object list dependency in 'cpumanf' is erroneous--have you tried removing that?
> Does that fix the problem, too (without the "ref_count <= 0" one-liner)?
> 
>        --SK

issues@scons said at 2007-04-19 10:28:03
>OK, I've got some extra info.
> 
>Sorting the Alias did not fix the problem. Reworking the Tool so that the .map
>is treated as a side-effect and not a first-class output of the linker did,
>however. (TBH, this is how it should have been from the first case but I didn't
>want to change inherited code without a reason.)
> 
>Anyway, I've created a snapshot of the failing structure on my machine, so I can
>still reproduce the bug.
> 
>Anyway, I thought that I would dig a little to try to help, so I put some prints
>in the Taskmaster to see where reference counts are set and decremented and I
>get the following:
> 
>...\cpumanf\build\c6416\cpumanf.out is the file with the problem.
> 
>It's reference count initially gets set to 3 (cpumanf_pre_cksum.out,
>cpumanf_pre_cksum.map, addcheck.exe) by the line "node.ref_count =
>len(set(not_started))"
> 
>Later, some time after the addcheck.exe ref_count is decremented by 1 as a
>parent of addcheck.obj (presumably to zero), the line "node.ref_count =
>len(set(not_built))" sets the ref_count of cpumanf.out to 2
>(cpumanf_pre_cksum.out, cpumanf_pre_cksum.map).
> 
>The reference count is then decremented by 1 as a parent of addcheck.exe.
> 
>After the link is performed, producing the two remaining dependencies, the
>ref_count is decremented by 2, resulting in -1.
> 
>Hope that is of some help - let me know if there is more that I can do,
> 
>Steve.
> 

stevenknight said at 2007-04-19 16:22:53
>Hi Steve--
> 
>Many thanks for the ref_count detective work, that's a great clue.  That gives
>me enough to conceptually see how it would happen, but I'm having a hard time
>nailing it down to the point where I can construct a consistent test case.  I've
>whipped together a script that should logically recreate the relationships
>between cpumanf.out, cpumanf_pre_cksum.{out,map} and addcheck.exe, and I see the
>reference counts behaving similar to how you're desribing it, but I can't
>actually trigger the re-initialization of the cpumanf.out ref_count to 2.  Of
>course, this may be timing related and extremely difficult (or impossible) to
>replicate consistently elsewhere, but I want to keep trying.
> 
>In the meantime, I'd like to throw a patch over the wall to you to test a
>grasping-at-straws theory about what might be causing the problem.  I'll attach
>it after this message, and if you could apply it to Taskmaster.py and give it a
>test on your broken snapshot, I'd appreciate it.  If it actually seems to fix
>the problem, at least I'll know my theory is in the ballpark.
> 
>Thanks,
> 
>        --SK
> 
> 
>        --SK

stevenknight said at 2007-04-19 16:26:15
>Created an attachment (id=179)
>*possible* fix for Taskmaster.py

issues@scons said at 2007-04-20 01:29:03
>Sorry, no - that doesn't fix the problem.
> 
>Cheers,
> 
>Steve.

stevenknight said at 2007-04-20 02:56:50
>Hi Steve--
> 
>Thanks for trying the Taskmaster.py patch.  Does it behave any differently if
>you try it with -j3 or -j5?  If it breaks when run with -j4 but works with other
>values, that would suggest that it's very specific to the order in which
>candidate Nodes are pulled off the stack on successive iterations through the
>loop in Job.py that controls how individual worker threads get their tasks.
> 
>Thanks,
> 
>        --SK

issues@scons said at 2007-04-20 08:36:55
>Well, I've tried the different numbers fo threads and it seems 100% reproducible
>with -j4 and -j5 but only about 50% with -j3... Go figure.
> 
>S.

stevenknight said at 2007-04-20 09:02:05
>Excellent clue, many thanks.  That does suggest that it's due to the interaction
>between the Job module and the Taskmaster, and that you happened to have
>stumbled on a way (lucky you!) to put exactly the wrong list of Nodes with the
>wrong states on the candidate list so that one of the Job iterations trips this
>condition.  Now I have to figure out how to characterize it well enough to
>reproduce it and then isolate an actual fix...  I'll let you know if I need
>still more info.
> 
>Thanks,
> 
>       --SK

issues@scons said at 2007-04-20 09:20:43
>Created an attachment (id=180)
>Instrumented successful build with -j 3

issues@scons said at 2007-04-20 09:21:41
>Created an attachment (id=181)
>Instrumented unsuccessful build with -j 4

issues@scons said at 2007-04-20 09:27:48
>Hopefully I'll get this in before our servers go down for some weekend
>electrical maintanence.
> 
>I've attached logs of a successful build with -j 3 and an unsuccessful with -j
>4. I've instrumented Job.py, when a new task is allocated to a thread, and
>Taskmaster in the places where the ref_count is set (both places) and decremented.
> 
>Quickly, it looks like the problem is where the ref_count gets decremented due
>to "building" addcheck relative to the setting of the ref_count from not_built.
>Is there a race hazard where the not_build list is updated before the ref_count
>is updated, and another thread gets in in the middle?
> 
>Anyway, hope that helps. I'll check in again on Monday and will give whatever
>additional help I can then.
> 
>Cheers and have a good weekend,
> 
>Steve.

issues@scons said at 2007-04-24 04:41:35
>Created an attachment (id=184)
>actual fix for the problem, to be released in 0.97 (or 0.96.97)

stevenknight said at 2007-04-24 04:43:54
>Okay, I seem to have this one nailed.  I kicking off a build and regression test
>run overnight to make sure, and everything looks good, so I'm going ahead with
>integration.
> 
>That was my ref_count.patch file just attached, I forgot to log in first...
> 
>The problem had to do with the two cpumanf_pre_cksum.{out,map} files being built
>simultaneously.  If one of those Nodes were visited before addcheck.exe finished
>building, then the reference count would get set to two (for the two files),
>then addcheck.exe would finish and decrement the ref count to 1, and then it
>would get decremented to -1 when the two files finished.  The solution was to
>fix how the ref_count is adjusted when we're waiting for Nodes that haven't
>finished building.
> 
>Because it's a race condition, I couldn't think of a good way to force it to
>happen every time without actually going in and instrumenting SCons itself.  The
>test script I came up with seems to fail for me about two thirds of the time on
>a sufficiently fast system, and is at least better than nothing.
> 
>Many, many thanks for the information you supplied.  It was crucial in tracking
>this down--in fact, I used your file names in the test script.

issues@scons said at 2007-04-25 08:12:17
>SK,
> 
>I've applied the patch and it has fixed the problem for me. 
> 
>Great work! Thanks!
> 
>Steve.

An anonymous user attached [j4pass1.txt](https://github.com/SCons/tigris-issue-attachments/blob/master/1636/174/j4pass1.txt) at 2007-04-17 08:15:04.
>log of "scons cpumanf -j 4 --tree=all --taskmastertrace=-" for the attempt

An anonymous user attached [j4pass2.txt](https://github.com/SCons/tigris-issue-attachments/blob/master/1636/175/j4pass2.txt) at 2007-04-17 08:16:36.
>log of "scons cpumanf -j 4 --tree=all --taskmastertrace=-" for the second attempt

An anonymous user attached [j1.txt](https://github.com/SCons/tigris-issue-attachments/blob/master/1636/176/j1.txt) at 2007-04-17 08:17:56.
>log of "scons cpumanf --tree=all --taskmastertrace=-" (which succeeds in one attempt)

An anonymous user attached [j3verbose.txt](https://github.com/SCons/tigris-issue-attachments/blob/master/1636/180/j3verbose.txt) at 2007-04-20 09:20:43.
>Instrumented successful build with -j 3

An anonymous user attached [j4verbose.txt](https://github.com/SCons/tigris-issue-attachments/blob/master/1636/181/j4verbose.txt) at 2007-04-20 09:21:41.
>Instrumented unsuccessful build with -j 4

An anonymous user attached [ref_count.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/1636/184/ref_count.patch) at 2007-04-24 04:41:35.
>actual fix for the problem, to be released in 0.97 (or 0.96.97)

stevenknight attached [maybe_fix.patch](https://github.com/SCons/tigris-issue-attachments/blob/master/1636/179/maybe_fix.patch) at 2008-03-25 11:08:15.
>*possible* fix for Taskmaster.py
